File: crates/hook/src/context_usage.rs
//! Claude status-line bridge. Only numeric context metadata leaves this process.
//! Sequence allocation happens before stdin consumption; network arrival order
//! and wall-clock adjustments cannot replace a newer observation with an old one.

use std::fs::{self, OpenOptions};
use std::io::{self, Read, Seek, SeekFrom, Write};
use std::path::Path;
use std::time::{Duration, Instant};

use crate::{
    endpoint::{env_var, is_valid_node_id},
    http::Request,
    Session, MAX_PAYLOAD_BYTES,
};
use serde_json::{json, Value};

const MAX_COUNT: u64 = 9_007_199_254_740_991;

pub fn run() -> i32 {
    // Always drain stdin, including outside Armadra and on any local failure.
    let binding = load_binding();
    let mut bytes = Vec::new();
    let mut input = io::stdin().lock();
    let read = (&mut input)
        .take(MAX_PAYLOAD_BYTES as u64 + 1)
        .read_to_end(&mut bytes);
    let _ = io::copy(&mut input, &mut io::sink());
    if read.is_err() || bytes.len() > MAX_PAYLOAD_BYTES {
        return 0;
    }
    let Some((session, session_id, generation, revision)) = binding else {
        return 0;
    };
    let Some(data) = serde_json::from_slice::<Value>(&bytes)
        .ok()
        .and_then(|input| filter_data(&input))
    else {
        return 0;
    };
    let body = json!({ "nodeId": session.node_id, "version": 1,
        "payload": { "armadraContextUsage": {
            "sessionId": session_id, "generation": generation,
            "sourceRevision": revision.to_string(), "data": data,
        }} });
    if let Ok(bytes) = serde_json::to_vec(&body) {
        let _ = session.send(|session, candidate| {
            Request::post_json(
                "/hook/claude",
                session.headers_for(candidate),
                bytes.clone(),
            )
        });
    }
    0
}

pub(crate) fn load_binding() -> Option<(Session, String, u64, u64)> {
    let session = Session::load().ok()?;
    // The sequence file has to live at one fixed path for the life of a
    // generation, so it is anchored to the first (preferred) candidate rather
    // than whichever one a later `send` happens to succeed through — the
    // ordering `discover_candidates` produces is itself stable across calls.
    let primary = session.candidates.first()?;
    primary.node_token(&session.node_id)?;
    let session_id = env_var("ARMADRA_SESSION_ID").filter(|id| is_valid_node_id(id))?;
    let generation = env_var("ARMADRA_SESSION_GENERATION")?
        .parse::<u64>()
        .ok()
        .filter(|value| *value <= MAX_COUNT)?;
    let parent = primary.path.parent()?;
    let directory = parent.join("context-sequences");
    let metadata = fs::symlink_metadata(&directory).ok()?;
    if !metadata.is_dir() || metadata.file_type().is_symlink() {
        return None;
    }
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        if metadata.permissions().mode() & 0o077 != 0 {
            return None;
        }
    }
    let revision = next_revision(&directory.join(format!("{session_id}-{generation}.seq"))).ok()?;
    Some((session, session_id, generation, revision))
}

/// OS locks release on process termination. Corrupt files are never reset to
/// zero: doing so could make delayed old reports look newer after a crash.
pub fn next_revision(path: &Path) -> io::Result<u64> {
    if let Ok(metadata) = fs::symlink_metadata(path) {
        if !metadata.is_file() || metadata.file_type().is_symlink() {
            return Err(io::Error::other("invalid context sequence file"));
        }
    }
    let mut options = OpenOptions::new();
    // Runtime initializes this exact generation once before spawning the PTY.
    // A removed file is not permission to reset an active generation to zero.
    options.read(true).write(true);
    #[cfg(unix)]
    {
        use std::os::unix::fs::OpenOptionsExt;
        options.mode(0o600);
    }
    let mut file = options.open(path)?;
    // Bounded, so a hook never hangs a CLI on a stuck holder — but wide
    // enough for the holders that are merely slow: each one reads, writes and
    // syncs the file, and eight of them on a loaded host add up to more than
    // the first bound of 150 ms.
    let deadline = Instant::now() + Duration::from_millis(500);
    loop {
        match file.try_lock() {
            Ok(()) => break,
            Err(std::fs::TryLockError::WouldBlock) if Instant::now() < deadline => {
                std::thread::sleep(Duration::from_millis(5))
            }
            // Distinguishable, so a caller that can afford to try again knows
            // this was contention and not a broken file.
            Err(std::fs::TryLockError::WouldBlock) => {
                return Err(io::Error::new(
                    io::ErrorKind::WouldBlock,
                    "context sequence is held by another hook",
                ));
            }
            Err(std::fs::TryLockError::Error(error)) => return Err(error),
        }
    }
    let length = file.metadata()?.len();
    let previous = match length {
        16 => {
            let mut bytes = [0u8; 16];
            file.read_exact(&mut bytes)?;
            let count = u64::from_be_bytes(bytes[..8].try_into().unwrap());
            let inverse = u64::from_be_bytes(bytes[8..].try_into().unwrap());
            if count != !inverse {
                return Err(io::Error::other("corrupt context sequence"));
            }
            count
        }
        _ => return Err(io::Error::other("corrupt context sequence")),
    };
    let next = previous
        .checked_add(1)
        .ok_or_else(|| io::Error::other("context sequence exhausted"))?;
    file.seek(SeekFrom::Start(0))?;
    file.write_all(&next.to_be_bytes())?;
    file.write_all(&(!next).to_be_bytes())?;
    file.sync_data()?;
    Ok(next)
}

/// Deliberately excludes transcript paths, workspace paths, costs, tools,
/// prompts and account limits. Missing current usage is forwarded as null.
pub fn filter_data(input: &Value) -> Option<Value> {
    let text = |value: &Value, key: &str| -> Option<String> {
        let text = value.get(key)?.as_str()?;
        (!text.is_empty() && text.len() <= 200 && !text.chars().any(char::is_control))
            .then(|| text.into())
    };
    let session = text(input, "session_id")?;
    let model = text(input.get("model")?, "id")?;
    let window = input.get("context_window")?;
    let current = window.get("current_usage").unwrap_or(&Value::Null);
    let usage = if current.is_null() {
        Value::Null
    } else {
        let count = |key| {
            current
                .get(key)?
                .as_u64()
                .filter(|value| *value <= MAX_COUNT)
        };
        json!({ "input_tokens": count("input_tokens")?,
            "cache_creation_input_tokens": count("cache_creation_input_tokens")?,
            "cache_read_input_tokens": count("cache_read_input_tokens")? })
    };
    let mut filtered = json!({ "session_id": session, "model": { "id": model },
        "context_window": { "context_window_size": window.get("context_window_size").and_then(Value::as_u64).filter(|value| *value > 0 && *value <= MAX_COUNT),
            "current_usage": usage } });
    if window.get("current_usage").is_none() {
        filtered["context_window"]
            .as_object_mut()?
            .remove("current_usage");
    }
    Some(filtered)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    #[test]
    fn only_current_context_metadata_is_forwarded() {
        let input = json!({ "session_id":"s", "model":{"id":"m", "secret":"hidden"},
            "transcript_path":"private", "cost":{"total":500}, "prompt":"private",
            "context_window":{"total_input_tokens":99999999, "context_window_size":200000,
            "current_usage":{"input_tokens":100,"cache_creation_input_tokens":20,"cache_read_input_tokens":30,"output_tokens":90}}});
        let output = filter_data(&input).unwrap();
        assert_eq!(
            output["context_window"]["current_usage"]
                .as_object()
                .unwrap()
                .len(),
            3
        );
        assert!(!output.to_string().contains("private"));
        assert!(!output.to_string().contains("total_input"));
        assert!(!output.to_string().contains("output_tokens"));
        let mut compact = input;
        compact["context_window"]["current_usage"] = Value::Null;
        assert!(filter_data(&compact).unwrap()["context_window"]["current_usage"].is_null());
    }
    #[test]
    fn revisions_are_monotonic_across_parallel_invocations() {
        let directory = tempfile::tempdir().unwrap();
        let path = directory.path().join("session.seq");
        fs::write(&path, [0u64.to_be_bytes(), u64::MAX.to_be_bytes()].concat()).unwrap();
        // The subject here is monotonicity, not the bounded wait (which has
        // its own test below): eight holders each sync the file, and on a
        // loaded host the last of them can legitimately be told to try again.
        let mut values = std::thread::scope(|scope| {
            (0..8)
                .map(|_| {
                    let path = &path;
                    scope.spawn(move || loop {
                        match next_revision(path) {
                            Ok(value) => break value,
                            Err(error) if error.kind() == io::ErrorKind::WouldBlock => continue,
                            Err(error) => panic!("{error}"),
                        }
                    })
                })
                .collect::<Vec<_>>()
                .into_iter()
                .map(|join| join.join().unwrap())
                .collect::<Vec<_>>()
        });
        values.sort();
        assert_eq!(values, (1..=8).collect::<Vec<_>>());
        assert_eq!(next_revision(&path).unwrap(), 9);
    }
    #[test]
    fn broken_sequence_is_not_reinitialized_and_lock_wait_is_bounded() {
        let directory = tempfile::tempdir().unwrap();
        let path = directory.path().join("session.seq");
        assert!(next_revision(&path).is_err());
        fs::write(&path, [1, 2, 3]).unwrap();
        assert!(next_revision(&path).is_err());
        assert_eq!(fs::read(&path).unwrap(), [1, 2, 3]);
        let file = File::options().read(true).write(true).open(&path).unwrap();
        file.lock().unwrap();
        let start = Instant::now();
        assert!(next_revision(&path).is_err());
        assert!(start.elapsed() < Duration::from_secs(1));
    }
}

File: apps/desktop/src/cli/armadra-hook/context-usage.ts
/**
 * Claude status-line bridge. Only numeric context metadata leaves this
 * process. Sequence allocation happens before stdin consumption; network
 * arrival order and wall-clock adjustments cannot replace a newer observation
 * with an old one.
 */

import * as fs from "node:fs";
import * as path from "node:path";

import { envVar, isValidNodeId, nodeToken } from "./endpoint.js";
import {
  asNumber,
  asObject,
  asString,
  canonicalJsonBytes,
  parseJson,
} from "./json.js";
import type { JsonValue } from "./json.js";
import { postJsonRequest } from "./http.js";
import { headersFor, loadSession, send } from "./session.js";
import type { Session } from "./session.js";
import { MAX_PAYLOAD_BYTES } from "./usage.js";

const MAX_COUNT = 9_007_199_254_740_991n;

/** How long a caller will wait for another hook to release the sequence. */
const LOCK_DEADLINE_MS = 500;
/**
 * A lock is held for a single read-modify-write, so one older than this was
 * left behind by a process that died mid-update and must not wedge the next
 * hook. Node has no `flock`, which is what the Rust client uses; an exclusive
 * `O_EXCL` sidecar with a staleness bound is the portable equivalent.
 */
const LOCK_STALE_MS = 5_000;

export interface Binding {
  session: Session;
  sessionId: string;
  generation: number;
  revision: bigint;
}

/**
 * What this command prints, which is what Claude draws as the status line.
 *
 * Only the node's name, and only when the node has one: the status line is a
 * single line of the user's own screen, so it gets the one fact the model and
 * the user both need and nothing else. An unnamed node prints nothing, which
 * is the same blank line this command has always produced.
 *
 * This is **our** status line — `hook/install/claude.ts` writes it only when
 * the user has none of their own, and a foreign one is never wrapped. So
 * appending here cannot touch a line somebody else wrote.
 */
export function statusLine(name: string | undefined): string {
  const handle = name?.trim() ?? "";
  // Re-validated rather than trusted: this is an environment variable, and a
  // control character in it would move the cursor rather than read as a name.
  if (!/^[a-z0-9][a-z0-9_-]{0,23}$/.test(handle)) return "";
  return `@${handle}`;
}

export async function run(): Promise<number> {
  const line = statusLine(envVar("ARMADRA_NODE_NAME"));
  if (line !== "") process.stdout.write(`${line}\n`);
  // Always drain stdin, including outside Armadra and on any local failure.
  const binding = loadBinding();
  const chunks: Buffer[] = [];
  let total = 0;
  try {
    for await (const chunk of process.stdin) {
      const buffer = chunk as Buffer;
      total += buffer.length;
      if (total <= MAX_PAYLOAD_BYTES + 1) chunks.push(buffer);
    }
  } catch {
    return 0;
  }
  if (total > MAX_PAYLOAD_BYTES) return 0;
  if (binding === undefined) return 0;
  const parsed = tryParse(Buffer.concat(chunks).toString("utf8"));
  const data = parsed === undefined ? undefined : filterData(parsed);
  if (data === undefined) return 0;
  const body = canonicalJsonBytes({
    nodeId: binding.session.nodeId,
    version: 1,
    payload: {
      armadraContextUsage: {
        sessionId: binding.sessionId,
        generation: binding.generation,
        sourceRevision: binding.revision.toString(),
        data,
      },
    },
  });
  await send(binding.session, (current, candidate) =>
    postJsonRequest("/hook/claude", headersFor(current, candidate), body),
  );
  return 0;
}

function tryParse(text: string): JsonValue | undefined {
  try {
    return parseJson(text);
  } catch {
    return undefined;
  }
}

export function loadBinding(): Binding | undefined {
  const loaded = loadSession();
  if ("error" in loaded) return undefined;
  const session = loaded.ok;
  // The sequence file has to live at one fixed path for the life of a
  // generation, so it is anchored to the first (preferred) candidate rather
  // than whichever one a later `send` happens to succeed through.
  const primary = session.candidates[0];
  if (primary === undefined) return undefined;
  if (nodeToken(primary, session.nodeId) === undefined) return undefined;
  const sessionId = envVar("ARMADRA_SESSION_ID");
  if (sessionId === undefined || !isValidNodeId(sessionId)) return undefined;
  const rawGeneration = envVar("ARMADRA_SESSION_GENERATION");
  if (rawGeneration === undefined || !/^\d+$/.test(rawGeneration))
    return undefined;
  const generation = BigInt(rawGeneration);
  if (generation > MAX_COUNT) return undefined;

  const directory = path.join(path.dirname(primary.path), "context-sequences");
  let metadata: fs.Stats;
  try {
    metadata = fs.lstatSync(directory);
  } catch {
    return undefined;
  }
  if (!metadata.isDirectory() || metadata.isSymbolicLink()) return undefined;
  if (process.platform !== "win32" && (metadata.mode & 0o077) !== 0)
    return undefined;

  try {
    const revision = nextRevision(
      path.join(directory, `${sessionId}-${generation}.seq`),
    );
    return { session, sessionId, generation: Number(generation), revision };
  } catch {
    return undefined;
  }
}

/**
 * Corrupt files are never reset to zero: doing so could make delayed old
 * reports look newer after a crash. The runtime initialises this exact
 * generation once before spawning the PTY, so a removed file is not permission
 * to reset an active generation either — the file is opened, never created.
 */
export function nextRevision(file: string): bigint {
  let stats: fs.Stats | undefined;
  try {
    stats = fs.lstatSync(file);
  } catch {
    stats = undefined;
  }
  if (stats !== undefined && (!stats.isFile() || stats.isSymbolicLink())) {
    throw new Error("invalid context sequence file");
  }
  const release = acquireLock(file);
  try {
    const handle = fs.openSync(file, "r+");
    try {
      if (fs.fstatSync(handle).size !== 16)
        throw new Error("corrupt context sequence");
      const buffer = Buffer.alloc(16);
      fs.readSync(handle, buffer, 0, 16, 0);
      const count = buffer.readBigUInt64BE(0);
      const inverse = buffer.readBigUInt64BE(8);
      if (count !== (~inverse & 0xffff_ffff_ffff_ffffn)) {
        throw new Error("corrupt context sequence");
      }
      const next = count + 1n;
      if (next > 0xffff_ffff_ffff_ffffn)
        throw new Error("context sequence exhausted");
      const out = Buffer.alloc(16);
      out.writeBigUInt64BE(next, 0);
      out.writeBigUInt64BE(~next & 0xffff_ffff_ffff_ffffn, 8);
      fs.writeSync(handle, out, 0, 16, 0);
      fs.fdatasyncSync(handle);
      return next;
    } finally {
      fs.closeSync(handle);
    }
  } finally {
    release();
  }
}

/**
 * A bounded exclusive lock, so a hook never hangs a CLI on a stuck holder —
 * but wide enough for the holders that are merely slow.
 */
function acquireLock(file: string): () => void {
  const lockPath = `${file}.lock`;
  const deadline = Date.now() + LOCK_DEADLINE_MS;
  for (;;) {
    try {
      fs.closeSync(fs.openSync(lockPath, "wx", 0o600));
      return () => {
        try {
          fs.rmSync(lockPath, { force: true });
        } catch {
          // Another holder already stole it; nothing left to release.
        }
      };
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "EEXIST") throw error;
    }
    try {
      if (Date.now() - fs.statSync(lockPath).mtimeMs > LOCK_STALE_MS) {
        fs.rmSync(lockPath, { force: true });
        continue;
      }
    } catch {
      continue;
    }
    if (Date.now() >= deadline) {
      // Distinguishable, so a caller that can afford to try again knows this
      // was contention and not a broken file.
      const error = new Error(
        "context sequence is held by another hook",
      ) as NodeJS.ErrnoException;
      error.code = "EWOULDBLOCK";
      throw error;
    }
    sleepBriefly();
  }
}

/** A 5 ms spin without an event-loop turn; `nextRevision` is synchronous. */
function sleepBriefly(): void {
  const until = Date.now() + 5;
  while (Date.now() < until) {
    // Busy wait: the hold time is microseconds, so this never runs long.
  }
}

/**
 * Deliberately excludes transcript paths, workspace paths, costs, tools,
 * prompts and account limits. Missing current usage is forwarded as null.
 */
export function filterData(input: JsonValue): JsonValue | undefined {
  const root = asObject(input);
  if (root === undefined) return undefined;
  const text = (
    value: Record<string, JsonValue> | undefined,
    key: string,
  ): string | undefined => {
    const field = asString(value?.[key]);
    if (field === undefined || field === "") return undefined;
    if (Buffer.byteLength(field, "utf8") > 200) return undefined;
    // `char::is_control` on the Rust side: the Unicode Cc category.
    if (/[\u0000-\u001f\u007f-\u009f]/.test(field)) return undefined;
    return field;
  };
  const sessionId = text(root, "session_id");
  if (sessionId === undefined) return undefined;
  const model = text(asObject(root["model"]), "id");
  if (model === undefined) return undefined;
  const window = asObject(root["context_window"]);
  if (window === undefined) return undefined;
  const hasCurrent = "current_usage" in window;
  const current = asObject(window["current_usage"]);

  let usage: JsonValue;
  if (!hasCurrent || window["current_usage"] === null) {
    usage = null;
  } else {
    if (current === undefined) return undefined;
    const count = (key: string): number | undefined => {
      const value = asNumber(current[key]);
      if (value === undefined || !Number.isInteger(value) || value < 0)
        return undefined;
      return BigInt(value) <= MAX_COUNT ? value : undefined;
    };
    const input_tokens = count("input_tokens");
    const cache_creation_input_tokens = count("cache_creation_input_tokens");
    const cache_read_input_tokens = count("cache_read_input_tokens");
    if (
      input_tokens === undefined ||
      cache_creation_input_tokens === undefined ||
      cache_read_input_tokens === undefined
    ) {
      return undefined;
    }
    usage = {
      input_tokens,
      cache_creation_input_tokens,
      cache_read_input_tokens,
    };
  }

  const size = asNumber(window["context_window_size"]);
  const contextWindowSize =
    size !== undefined &&
    Number.isInteger(size) &&
    size > 0 &&
    BigInt(size) <= MAX_COUNT
      ? size
      : null;
  const contextWindow: Record<string, JsonValue> = {
    context_window_size: contextWindowSize,
  };
  if (hasCurrent) contextWindow["current_usage"] = usage;
  return {
    session_id: sessionId,
    model: { id: model },
    context_window: contextWindow,
  };
}
